src/sourmash/tax/tax_utils.py

```diff
@@ -36,7 +39,272 @@
 
 # import lca utils as needed for now
 from sourmash.lca import lca_utils
-from sourmash.lca.lca_utils import (LineagePair, taxlist, display_lineage, pop_to_rank)
+from sourmash.lca.lca_utils import (taxlist, display_lineage, pop_to_rank)
+
+class LineagePair(NamedTuple):
+    rank: str
+    name: str = None
+    taxid: int = None
+
+@dataclass(frozen=True, order=True)
+class BaseLineageInfo:
+    """BaseLineageInfo Class, defining a set of functions that can be used to handle
+       hierarchical set of LineagePairs. """
+    # need to set compare=False for any mutable type to keep this class hashable
+    ranks: tuple() # require ranks
+    lineage: tuple = field(default=()) #tuple of LineagePairs
+    lineage_str: str = field(default=None, compare=False) # ';'- or ','-separated str of lineage names
+    lineage_dict: dict = field(default=None, compare=False) # dict of rank: name
+
+    def __post_init__(self):
+        "Initialize according to passed values"
+        # ranks must be tuple for hashability
+        if isinstance(self.ranks, list):
+            object.__setattr__(self, "ranks", tuple(self.ranks))
+        if self.lineage:
+            self._init_from_lineage_tuples()
+        elif self.lineage_str is not None:
+            self._init_from_lineage_str()
+        elif self.lineage_dict is not None:
+            self._init_from_lineage_dict()
+        else:
+            self._init_empty()
+
+    def __eq__(self, other):
+        if other == (): # just handy: if comparing to a null tuple, don't try to find it's lineage before returning False
+            return False
+        return all([self.ranks == other.ranks and self.lineage==other.lineage])
+
+    @property
+    def taxlist(self):
+        return self.ranks
+
+    @property
+    def ascending_taxlist(self):
+        return self.ranks[::-1]
+
+    @property
+    def lowest_rank(self):
+        if not self.filled_ranks:
+            return None
+        return self.filled_ranks[-1]
+
+    def rank_index(self, rank):
+        return self.ranks.index(rank)
+
+    @property
+    def filled_lineage(self):
+        """Return lineage down to lowest non-empty rank. Preserves missing ranks above."""
+        # Would we prefer this to be the default returned by lineage??
+        if not self.filled_ranks:
+            return ()
+        lowest_filled_rank_idx = self.rank_index(self.filled_ranks[-1])
+        return self.lineage[:lowest_filled_rank_idx+1]
+
+    @property
+    def lowest_lineage_name(self):
+        """Return the name of the lowest filled lineage"""
+        if not self.filled_ranks:
+            return ""
```

Comment:
To return to later: should this be `""` or `None`?

Reply:
this seems better-- changed